### PR TITLE
Copy the content from the underlying image into the newly created volume

### PR DESCRIPTION
**Pull request by vikas-goel**
Fixes: #10262

Signed-off-by: Vikas Goel <vikas.goel@gmail.com>

### CODE REVIEW

**Comment by vikas-goel** [comment]
@mheon , @rhatdan , @Luap99 , @TomSweeneyRedHat ,
I am facing two issues, one related to test case and another about the code change. I am posting them separately. I will appreciate any guidance.

**Comment by vikas-goel** [comment]
I am using `config/containers-volume.conf` in test case but the execution failing to find the file. I have tried creating the file runtime in tmp directory instead of static file, that doesn't help either.

```
[+0935s] Running: /var/tmp/go/src/github.com/containers/podman/bin/podman --storage-opt vfs.imagestore=/tmp/podman/imagecachedir --root /tmp/podman_test791424325/crio --runroot /tmp/podman_test791424325/crio-run --runtime crun --conmon /usr/bin/conmon --cni-config-dir /etc/cni/net.d --cgroup-manager cgroupfs --tmpdir /tmp/podman_test791424325 --events-backend file --storage-driver vfs create --name ctr-copy dataimg
[+0935s] time="2021-07-19T15:47:35Z" level=error msg="Error streaming contents of container c15a6d0d258da0180568a3a4a502dc19e9c9a8984b48bedeb52d292e1cfa3608 directory for volume copy-up: error in copier subprocess: time=\"2021-07-19T15:47:35Z\" level=error msg=\"finding config on system: CONTAINERS_CONF file: stat config/containers-volume.conf: no such file or directory\"\n"
[+0935s] Error: error copying up to volume d2e89f0d1df08c3999078ff13be71a999bf6d507468b348956ebe0ce360e9f71: error in copier subprocess: time="2021-07-19T15:47:35Z" level=error msg="finding config on system: CONTAINERS_CONF file: stat config/containers-volume.conf: no such file or directory"
```

File: libpod/runtime_ctr.go [hunk 1]
```diff
@@ -47,6 +47,32 @@ func (r *Runtime) NewContainer(ctx context.Context, rSpec *spec.Spec, options ..
...
 }
 
+func (r *Runtime) PrepareVolumeOnCreateContainer(ctx context.Context, ctr *Container) (retErr error) {
+	// Copy the content from the underlying image into the newly created
+	// volume if configured to do so.
+	if !r.config.Containers.PrepareVolumeOnCreate {
+		return nil
+	}
+
+	defer func() {
```
Comment (rhatdan) [inline libpod/runtime_ctr.go:57]: Since it's only the mountstorage that can fail, I don't think you need to cleanup storage at all.

File: libpod/runtime_ctr.go [hunk 2]
```diff
@@ -47,6 +47,32 @@ func (r *Runtime) NewContainer(ctx context.Context, rSpec *spec.Spec, options ..
...
+			}
+		}
+	}()
+
+	mountPoint, err := ctr.mountStorage()
+	if err == nil {
+		// Finish up mountStorage
+		ctr.state.Mounted = true
+		ctr.state.Mountpoint = mountPoint
+		logrus.Infof("Container %s storage %s mounted successfully", ctr.Name(), mountPoint)
```
Comment (rhatdan) [inline libpod/runtime_ctr.go:70]: This should be Debug, not Info.

**Comment by vikas-goel** [comment]
I am facing an issue where removing the container and volume created using `prepare_volume_on_create = true` show the following error message.
```
# ~/go/src/github.com/containers/podman/bin/podman rm -afv
Error: error removing container 58865407f8f7c1858704375a173a1cc93da4280cdb079aad44be536069540eff root filesystem: 2 errors occurred:
        * unlinkat /var/lib/containers/storage/overlay-containers/58865407f8f7c1858704375a173a1cc93da4280cdb079aad44be536069540eff/userdata/shm: device or resource busy
        * unlinkat /var/lib/containers/storage/overlay-containers/58865407f8f7c1858704375a173a1cc93da4280cdb079aad44be536069540eff/userdata/shm: device or resource busy
```

If I run `container init` and `container cleanup` after creating the container with `prepare_volume_on_create = true` option, then the error does not happen.

File: pkg/specgen/generate/container_create.go [hunk 3]
```diff
@@ -153,7 +153,17 @@ func MakeContainer(ctx context.Context, rt *libpod.Runtime, s *specgen.SpecGener
...
-	return rt.NewContainer(ctx, runtimeSpec, options...)
+
+	ctr, err := rt.NewContainer(ctx, runtimeSpec, options...)
+	if err != nil {
+		return ctr, err
+	}
+
+	// Copy the content from the underlying image into the newly created
+	// volume if configured to do so.
+	verr := rt.PrepareVolumeOnCreateContainer(ctx, ctr)
```
Comment (rhatdan) [inline pkg/specgen/generate/container_create.go:164]: Jusy use err.
I don't see the data being copied up in PrepareVolumeOnCreateContainer?

**Review by rhatdan — APPROVED** [review]

File: pkg/specgen/generate/container_create.go (hunk 3 above)
Comment (vikas-goel) [reply pkg/specgen/generate/container_create.go:164]: Using `err` was compilation error. I had to change that.
`PrepareVolumeOnCreateContainer()` calls `mountStorage()` that internally calls `mountNamedVolume()` which does the copy.

File: libpod/runtime_ctr.go (hunk 2 above)
Comment (vikas-goel) [reply libpod/runtime_ctr.go:70]: Sure, next commit will take care of that.

File: libpod/runtime_ctr.go (hunk 1 above)
Comment (vikas-goel) [reply libpod/runtime_ctr.go:57]: The `prepare()` function that calls `mountStorage()` does call `cleanupStorage()` afterwards. I am wondering whether not calling `cleanupStorage()` will result in any leak for the containers that are just created but never started/stopped.

File: libpod/runtime_ctr.go [hunk 4]
```diff
@@ -47,6 +47,32 @@ func (r *Runtime) NewContainer(ctx context.Context, rSpec *spec.Spec, options ..
 	return r.newContainer(ctx, rSpec, options...)
 }
 
+func (r *Runtime) PrepareVolumeOnCreateContainer(ctx context.Context, ctr *Container) (retErr error) {
```
Comment (mheon) [inline libpod/runtime_ctr.go:50]: Why did you move to a separate function for this, as opposed to keeping it part of container creation?

Comment (vikas-goel) [reply libpod/runtime_ctr.go:50]: When I had this code originally in `setupContainer()`, then `podman create` would hang indefinitely. Only a process kill would make the operation abort. I guess some of the deferred calls in the `setupContainer()` need to be executed before calling the volume copy.

So, I made the change to look it somewhat like `podman create` followed by `podman init` so as to follow the same business logic (code path).

**Comment by mheon** [comment]
@vikas-goel I've identified the problem. The issue is that `mountStorage()` is not setting `c.state.Mounted` which `unmountStorage()` checks before doing anything. After mounting storage you need to do a `c.state.Mounted = true` and then a `c.save()`.

**Comment by vikas-goel** [comment]
> @vikas-goel I've identified the problem. The issue is that `mountStorage()` is not setting `c.state.Mounted` which `unmountStorage()` checks before doing anything. After mounting storage you need to do a `c.state.Mounted = true` and then a `c.save()`.

That works. The missing save() was the reason. Thank you @mheon .

**Comment by TomSweeneyRedHat** [comment]
LGTM
assuming happy tests

**Comment by vikas-goel** [comment]
Assigned to @TomSweeneyRedHat for merge

File: libpod/runtime_ctr.go [hunk 5]
```diff
@@ -47,6 +47,33 @@ func (r *Runtime) NewContainer(ctx context.Context, rSpec *spec.Spec, options ..
...
+	// Copy the content from the underlying image into the newly created
+	// volume if configured to do so.
+	if !r.config.Containers.PrepareVolumeOnCreate {
+		return nil
+	}
+
+	defer func() {
+		if retErr != nil {
```
Comment (Luap99) [inline libpod/runtime_ctr.go:58]: There is no need to check for retErr, it is never used in this function.

File: libpod/runtime_ctr.go [hunk 6]
```diff
@@ -47,6 +47,33 @@ func (r *Runtime) NewContainer(ctx context.Context, rSpec *spec.Spec, options ..
 	return r.newContainer(ctx, rSpec, options...)
 }
 
+func (r *Runtime) PrepareVolumeOnCreateContainer(ctx context.Context, ctr *Container) (retErr error) {
```
Comment (Luap99) [inline libpod/runtime_ctr.go:50]:
```suggestion
func (r *Runtime) PrepareVolumeOnCreateContainer(ctx context.Context, ctr *Container) error {
```

File: libpod/runtime_ctr.go [hunk 7]
```diff
@@ -47,6 +47,33 @@ func (r *Runtime) NewContainer(ctx context.Context, rSpec *spec.Spec, options ..
...
+			}
+		}
+	}()
+
+	mountPoint, err := ctr.mountStorage()
+	if err == nil {
+		// Finish up mountStorage
+		ctr.state.Mounted = true
+		ctr.state.Mountpoint = mountPoint
+		ctr.save()
```
Comment (Luap99) [inline libpod/runtime_ctr.go:70]: You have to check the error from save().

**Comment by vikas-goel** [comment]
Not sure why some of the checks are cancelled and the status say "Some checks were not successful".

**Comment by vikas-goel** [comment]
Hi @TomSweeneyRedHat , could you help merge the PR is no other feedback?

File: pkg/specgen/generate/container_create.go (hunk 3 above)
Comment (rhatdan) [reply pkg/specgen/generate/container_create.go]: This should work.
```
err = rt.PrepareVolumeOnCreateContainer(ctx, ctr)
```

Comment (vikas-goel) [reply pkg/specgen/generate/container_create.go]: I changed the code to not worry about storing in variable. Please take another look.

**Comment by rhatdan** [comment]
Thanks @vikas-goel

**Comment by vikas-goel** [comment]
Thank you @rhatdan , @mheon , @TomSweeneyRedHat , @Luap99
